Add ability to select biospecimen template by type

Aryllen opened the pull request:
Fixes #495. Depends on #497.

Changes proposed in this pull request:

- New config options for having biospecimen templates by type and signaling to the UI to add the feature. The radiobutton options will reflect what the types are called in the template.
- Fix for drosophila logic. When switched to OAuth, lost the conditionalPanel logic for individual metadata. Added it back.
- Update to `gather_template_ids` to accept `biospecimen_type`.
- Update to `get_template_synIDs` to only gather synIDs and unlist the biospecimen templates if nested
- Updated docs
- Updated tests

Questions:

- Is "Specimen Type" a good name for this feature?
- What should the PEC config look like with this new addition?
- Should in vitro biospecimen be an option for all species in AD, except drosophila?

~~I will make a test version of the app on Friday.~~ [Test app here.]( https://www.synapse.org/#!Wiki:syn25976061/ENTITY) Note that this only has two assays as it's the default config.

Reviewing the code itself may be a bit confusing since this relies on a different branch that is also a PR, atm. I'll update this branch once I merge the other PR (hopefully later today).

Please confirm you've done the following (if applicable):

- [ ] If adding or altering a configuration value, opened a PR in [dccmonitor](https://github.com/Sage-Bionetworks/dccmonitor) to update the same configuration value or verified that the configuration option is irrelevant in dccmonitor.
- [x] Added tests for new functions or for new behavior in existing functions
- [ ] If adding a new exported function, added it to the reference section of `_pkgdown.yml`
- [x] If adding a new configuration option, documented it in `vignettes/customizing_dccvalidator.Rmd`
- [x] If adding or changing functionality, documented it under "dccvalidator (development version)" in `NEWS.md`

avanlinden commented:
> Is "Specimen Type" a good name for this feature?

I'd say maybe go with "Biospecimen Type", to explicitly link it to the metadata template. They will know which biospecimen template we've told them to use so hopefully that will make sense. Open to other opinions though!

Aryllen commented:
Based on conversation with @avanlinden, updated the name of the feature to "Biospecimen Type". I also updated the default config to show different ways to name the types (change the species to see the different options or how the feature goes away completely for drosophila). Whatever is first in the config list of biospecimen-type templates will be selected by default.

kelshmo commented on R/df-helpers.R:
```diff
@@ -59,10 +65,16 @@ gather_template_id_individual <- function(species) {
 ## gather_template_ids helper
-gather_template_id_biospecimen <- function(species) {
+gather_template_id_biospecimen <- function(species, biospecimen_type) {
   templates <- get_golem_config("templates")$biospecimen_templates
```
Having never worked with golem not sure this is relevant, but should templates be an input parameter and then `get_golem_config` called from the function `gather_template_id_biospecimen`? I'm thinking about how I integrate `config` with function arguments.

Aryllen replied:
I'm not exactly sure what you mean. `config` and `golem` (`get_golem_config()` calls `config::get()`) tend to assume a single list of options, not nested lists. However, it is a good point that I could grab the list of templates from the config in `gather_template_ids()` and pass it along, instead of grabbing it multiple times. I'll update this to be a little more efficient.

Aryllen replied:
Updated and made it a little nicer (IMO) for how I'm referencing specific lists of templates.

kelshmo replied:
Your updated code is spot on with my comment. Looks good!

kelshmo commented on R/df-helpers.R:
```diff
@@ -38,12 +38,18 @@ get_metadataType_indices <- function(data, meta_types) {
 #' @param species The species needed to specify the correct biospecimen
 #' or individual templates (default `NA`).
 #' @param assay The assay needed to specify the correct assay template.
+#' @param biospecimen_type The type of biospecimen template needed
```
Are accepted types "in_vitro" or "other"? What will happen if another value is passed?

Aryllen replied:
This accepts any values from the biospecimen template section of the config. The `biospecimen_type`s that will appear in the UI options will be pulled from the list of named biospecimen templates in the config. This adds extra flexibility for more types of biospecimen templates in the future, and gives you control over how the `biospecimen_type` appears in the app.

kelshmo replied:
fabulous.

kelshmo commented on R/metadata-template-dictionary.R:
```diff
@@ -307,8 +307,16 @@ generate_key_description <- function(annots) {
+  templates %>%
+    unname() %>%
+    unlist()
+}
```
```suggestion
}

```

kelshmo commented on vignettes/customizing-dccvalidator.Rmd:
```diff
@@ -124,11 +124,26 @@ To install the dccvalidator instead of forking the repository:
+  id, these should have a `properties` object listing all column names (a.k.a.
+  keys) that should be present in the metadata. The `manifest_template` should
+  have only one template. `individual_templates` and `biospecimen_templates`
+  should be broken up by species (the species must be included in the
```
I suggest revising the language to clarify "broken up by species": 
* `templates` (including `manifest_template`,......
     `individual_templates:`
                `<name species:> templates must be species-specific and the species must be named in a nested list under        `individual_templates`. 

It is difficult to capture the nested/indented relationship in this comment! Maybe instead of being explicit about the nested format, you could link to the appropriate line in the example config. Same thought goes for `biospecimen_templates`.

Aryllen replied:
This paragraph seemed cumbersome to me, as well! Thanks for the suggestion on how to make it better. I'll update it.

Aryllen replied:
Updated.

kelshmo reviewed (approved):
Approving, as the test app is working well! I included some comments of where the code/documentation could be made more explicit/clarified.

Aryllen commented:
Realized that I would need a biospecimenType annotation on the uploaded metadata files in order to make dccmonitor compatible. Updated to add this annotation on biospecimen metadata. Also made a few updates based on @kelshmo's feedback.

Aryllen commented:
Most of the checks have finished successfully after the latest merge of master. I am expecting the Windows check to fail, but that has to do with the setup of the .synapseConfig, not the package. Merging.

Note: dccmonitor should **not** be updated with the latest version of dccvalidator until I make dccmonitor compatible. Similarly, until dccmonitor is compatible and updated, the biospecimen metadata in dccmonitor will be checked against the main species-based biospecimen metadata; it will not check based on biospecimen type.
